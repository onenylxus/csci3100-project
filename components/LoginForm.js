// Import
import React from 'react';
import {
  Alert,
  Button,
  Text,
  TextInput,
  View,
  TouchableOpacity,
} from 'react-native';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import {
  faUser,
  faLock,
  faEyeSlash,
  faEye,
} from '@fortawesome/free-solid-svg-icons';
import { useNavigation } from '@react-navigation/native';
import Source from '../assets/source';
import Style from '../assets/style';

// Export login form
export default function LoginForm() {
  const navigation = useNavigation();

  const [username, setUsername] = React.useState('');
  const [password, setPassword] = React.useState('');
  const [visibility, setVisibility] = React.useState(true);

  let status = 0;

  async function submitData() {
    await fetch(`https://${Source.heroku}/login`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        username,
        password,
      }),
    })
      .then((res) => {
        status = res.status;
        return res;
      })
      .then((res) => res.json())
      .then((res) => {
        if (status === 200) {
          console.log(res.msg);
          navigation.navigate('Tabs', { username });
        } else if (status === 422) {
          console.log(res.error);
          switch (res.error) {
            case 'accountError':
              return Alert.alert(
                'Account does not exist',
                'We could not find an account with this username, please try again.',
                [
                  {
                    text: 'OK',
                    onPress: () => undefined,
                    style: 'destructive',
                  },
                ]
              );

            case 'passwordError':
              return Alert.alert(
                'Wrong password',
                'The password you provided is incorrect. Please try again.',
                [
                  {
                    text: 'OK',
                    onPress: () => undefined,
                    style: 'destructive',
                  },
                ]
              );

            default:
              return new Error();
          }
        }
      })
      .catch((err) => console.log(err));
  }

  async function confirmLogin() {
    if ([username, password].every((data) => data.length > 0)) {
      console.log(
        `Login request sent with username ${username} and password ${password}`
      );
      submitData();
    } else {
      return Alert.alert(
        'Some information are missing',
        'Some fields are missing. Fill out all fields and try again.',
        [
          {
            text: 'OK',
            onPress: () => undefined,
            style: 'destructive',
          },
        ]
      );
    }
  }

  return (
    <View>
      {/* Username */}
      <View style={Style.inputContainer}>
        <Text style={Style.sectionText}>Username:</Text>
        <View style={Style.SectionStyle}>
          <FontAwesomeIcon
            icon={faUser}
            size={15}
            style={{ marginHorizontal: 5, marginVertical: 12 }}
          />
          <TextInput
            placeholder="Username"
            style={{ width: 200 }}
            onChangeText={(text) => setUsername(text)}
          />
        </View>
      </View>
      {/* Password */}
      <View style={Style.inputContainer}>
        <Text style={Style.sectionText}>Password:</Text>
        <View style={Style.SectionStyle}>
          <FontAwesomeIcon
            icon={faLock}
            size={15}
            style={{ marginHorizontal: 5, marginVertical: 12 }}
          />
          <TextInput
            placeholder="Password"
            style={{ width: 175 }}
            onChangeText={(text) => setPassword(text)}
            secureTextEntry={visibility}
            clearTextOnFocus={false}
            enablesReturnKeyAutomatically
          />
          <TouchableOpacity onPress={() => setVisibility(!visibility)}>
            <FontAwesomeIcon
              icon={visibility ? faEyeSlash : faEye}
              size={15}
              style={{ marginHorizontal: 5, marginVertical: 12 }}
            />
          </TouchableOpacity>
        </View>
      </View>
      <Button title="Login" onPress={confirmLogin} />
    </View>
  );
}
