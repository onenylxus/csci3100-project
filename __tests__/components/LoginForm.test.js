// Import
import React from 'react';
import { Alert } from 'react-native';
import { fireEvent, render } from '@testing-library/react-native';

// Import target component
import AuthContext from '../../components/AuthContext';
import LoginForm from '../../components/LoginForm';

// Mock FontAwesome icons
jest.mock('@fortawesome/react-native-fontawesome', () => ({
  FontAwesomeIcon: '',
}));

// Mock authentication method
const AuthMethodMock = {
  login: jest.fn(),
  logout: jest.fn(),
  getUser: jest.fn(),
};

// Spy on alerts
jest.spyOn(Alert, 'alert');

// Run jest tests
describe('LoginForm', () => {
  let element;

  beforeEach(() => {
    element = render(
      <AuthContext.Provider value={AuthMethodMock}>
        <LoginForm />
      </AuthContext.Provider>
    );
  });

  it('changes username when type', () => {
    const username = element.getByTestId('username');

    fireEvent.changeText(username, 'testac');
    expect(username.props.value).toBe('testac');
  });

  it('changes password when type', () => {
    const password = element.getByTestId('password');

    fireEvent.changeText(password, 'testpw');
    expect(password.props.value).toBe('testpw');
  });

  it('toggles password visibility when eye button is pressed', () => {
    const eye = element.getByTestId('eye');
    const password = element.getByTestId('password');

    expect(password.props.secureTextEntry).toBe(true);
    fireEvent.press(eye);
    expect(password.props.secureTextEntry).toBe(false);
    fireEvent.press(eye);
    expect(password.props.secureTextEntry).toBe(true);
  });

  it('rejects login with empty username', () => {
    const password = element.getByTestId('password');
    const button = element.getByTestId('login');

    fireEvent.changeText(password, 'testpw');
    fireEvent.press(button);
    setTimeout(() => expect(Alert.alert).toHaveBeenCalled(), 5000);
  });

  it('rejects login with empty password', () => {
    const username = element.getByTestId('username');
    const button = element.getByTestId('login');

    fireEvent.changeText(username, 'testac');
    fireEvent.press(button);
    setTimeout(() => expect(Alert.alert).toHaveBeenCalled(), 5000);
  });

  it('rejects login with unregistered account', () => {
    const username = element.getByTestId('username');
    const password = element.getByTestId('password');
    const button = element.getByTestId('login');

    fireEvent.changeText(username, 'a');
    fireEvent.changeText(password, 'testpw');
    fireEvent.press(button);
    setTimeout(() => expect(Alert.alert).toHaveBeenCalled(), 5000);
  });

  it('rejects login with wrong password', () => {
    const username = element.getByTestId('username');
    const password = element.getByTestId('password');
    const button = element.getByTestId('login');

    fireEvent.changeText(username, 'testac');
    fireEvent.changeText(password, 'a');
    fireEvent.press(button);
    setTimeout(() => expect(Alert.alert).toHaveBeenCalled(), 5000);
  });

  it('logs in with correct data', () => {
    const username = element.getByTestId('username');
    const password = element.getByTestId('password');
    const button = element.getByTestId('login');

    fireEvent.changeText(username, 'testac');
    fireEvent.changeText(password, 'testpw');
    fireEvent.press(button);
    setTimeout(() => expect(AuthMethodMock.login).toHaveBeenCalled(), 5000);
  });

  it('matches snapshot', () => {
    expect(element.toJSON()).toMatchSnapshot();
  });
});
