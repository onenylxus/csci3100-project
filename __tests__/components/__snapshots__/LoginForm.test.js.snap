// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<View
  style={
    Object {
      "alignSelf": "center",
      "width": 400,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "alignSelf": "center",
        "backgroundColor": "#ffffff",
        "flexDirection": "column",
        "justifyContent": "center",
        "marginBottom": 16,
        "maxWidth": "80%",
        "minWidth": "80%",
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontFamily": "Roboto",
          "marginLeft": "2%",
        }
      }
    >
      Username:
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#000000",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "marginVertical": "5%",
          "maxWidth": "75%",
          "minWidth": "75%",
          "paddingHorizontal": "1%",
        }
      }
    >
      <
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f007",
              "M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z",
            ],
            "iconName": "user",
            "prefix": "fas",
          }
        }
        size={15}
        style={
          Object {
            "marginHorizontal": 5,
            "marginVertical": 12,
            "width": "50%",
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "maxWidth": "85%",
            "minWidth": "85%",
          }
        }
        testID="username"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      style={
        Object {
          "alignSelf": "flex-start",
          "fontFamily": "Roboto",
          "marginLeft": "2%",
        }
      }
    >
      Password:
    </Text>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#ffffff",
          "borderColor": "#000000",
          "borderRadius": 5,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 40,
          "marginVertical": "5%",
          "maxWidth": "75%",
          "minWidth": "75%",
          "paddingHorizontal": "1%",
        }
      }
    >
      <
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f023",
              "M400 224h-24v-72C376 68.2 307.8 0 224 0S72 68.2 72 152v72H48c-26.5 0-48 21.5-48 48v192c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V272c0-26.5-21.5-48-48-48zm-104 0H152v-72c0-39.7 32.3-72 72-72s72 32.3 72 72v72z",
            ],
            "iconName": "lock",
            "prefix": "fas",
          }
        }
        size={15}
        style={
          Object {
            "marginHorizontal": 5,
            "marginVertical": 12,
            "width": "50%",
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        clearTextOnFocus={false}
        enablesReturnKeyAutomatically={true}
        onChangeText={[Function]}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "maxWidth": "77%",
            "minWidth": "77%",
          }
        }
        testID="password"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="eye"
      >
        <
          icon={
            Object {
              "icon": Array [
                640,
                512,
                Array [],
                "f070",
                "M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z",
              ],
              "iconName": "eye-slash",
              "prefix": "fas",
            }
          }
          size={15}
          style={
            Object {
              "marginHorizontal": 5,
              "marginVertical": 12,
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "marginHorizontal": "40%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="login"
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Login
        </Text>
      </View>
    </View>
  </View>
</View>
`;
